sabnzbd/downloader.py

```diff
@@ -545,29 +547,33 @@ def decode(self, article, raw_data: Optional[bytearray] = None, raw_data_size: O
 
         # See if we need to delay because the queues are full
         logged_counter = 0
-        decoder_full = sabnzbd.Decoder.queue_full()
-        assembler_full = sabnzbd.Assembler.queue_full()
-        while not self.shutdown and (decoder_full or assembler_full):
-            # Only log/update once every second, to not waste any CPU-cycles
-            if not logged_counter % 10:
-                # Make sure the BPS-meter is updated
-                sabnzbd.BPSMeter.update()
-
-                # Update who is delaying us
-                sabnzbd.BPSMeter.delayed_decoder += int(decoder_full)
-                sabnzbd.BPSMeter.delayed_assembler += int(assembler_full)
-                logging.debug(
-                    "Delayed - %d seconds - Decoder queue: %d - Assembler queue: %d",
-                    logged_counter / 10,
-                    sabnzbd.Decoder.decoder_queue.qsize(),
-                    sabnzbd.Assembler.queue.qsize(),
-                )
 
-            # Wait and update the queue sizes
-            time.sleep(0.1)
-            logged_counter += 1
-            decoder_full = sabnzbd.Decoder.queue_full()
-            assembler_full = sabnzbd.Assembler.queue_full()
+        decoder_level = sabnzbd.Decoder.queue_level()
+        assembler_level = sabnzbd.Assembler.queue_level()
+
+        # Sleep for an increasing amount of time, depending on queue sizes.
+        if decoder_level > _QSLOWDOWN_LEVEL or assembler_level > _QSLOWDOWN_LEVEL:
+            time.sleep((decoder_level + assembler_level - _QSLOWDOWN_LEVEL) / 2)
```

Comment:
The `/2` is arbitrary, right? Why did you chose it?

Reply:
Mostly because a maximum sleep of 0.2 seconds (for one full queue) seemed fair. If it triggers every time a part completes it will slow down the speed to 3.75 MB/s before it's full and the absolute limit kicks in. For the assembler queue the minimum sleep will be 0.05 seconds with a limit of 10 elements. It won't slow it down very significantly if it's only needed once in a while to make the processing catch up, but at the same time it's not so short that it's insignificant if it happens on every call. It hopefully gives the assembler queue some granularity in the slowing down even though it has a fairly low max limit.

If it's higher then it should probably be split up in multiple shorter sleeps with a check for shutdown and other things in between. I wanted to avoid that for simplicity. If it's much smaller then the cases where it's not enough will increase.

If it should be changed it's probably downwards a bit (for a shorter sleep). Maybe /4. I think anything less than that would be mostly useless for the assembler queue.